In [1]:
#instalando e importanto librerías necesarias para el Proyecto
!pip install geopandas
!pip install folium
import numpy as np 
import pandas as pd 
import geopandas as gpd

import matplotlib.pyplot as plt

import folium
from folium import Choropleth, Circle, Marker
from folium.plugins import HeatMap, MarkerCluster


[notice] A new release of pip available: 22.2.2 -> 22.3
[notice] To update, run: python.exe -m pip install --upgrade pip

[notice] A new release of pip available: 22.2.2 -> 22.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
#Importamos el archivo train como data y  test coomo data_test
data = pd.read_csv(r"C:\Users\Juan\Desktop\PROYECTO INDIVIDUAL\PROYECTO INDIVIDAL 02\Datathon\properties_colombia_train.csv", encoding="utf-8")
data_test = pd.read_csv(r"C:\Users\Juan\Desktop\PROYECTO INDIVIDUAL\PROYECTO INDIVIDAL 02\Datathon\properties_colombia_test.csv")

In [4]:
#Mostrando las 3 primeras filas de data_test
data_test.head(3)

Unnamed: 0                        id    ad_type  start_date    end_date  \
0      852569  pCiVu6rxpBq9dCZfYCXXgA==  Propiedad  2020-09-04  2020-09-12   
1      121379  L0XSQ7mF1TWb1lHRg4n2+A==  Propiedad  2020-10-20  2020-10-20   
2      119504  FbWRxZOMNoUzVOLXgkum1A==  Propiedad  2020-10-20  2020-10-22   

   created_on       lat        lon        l1            l2  ... bathrooms  \
0  2020-09-04  4.722000 -74.075000  Colombia  Cundinamarca  ...       3.0   
1  2020-10-20       NaN        NaN  Colombia     Antioquia  ...       3.0   
2  2020-10-20  6.259253 -75.628381  Colombia     Antioquia  ...       1.0   

  surface_total surface_covered currency  price_period  \
0         660.0           350.0      COP           NaN   
1           NaN             NaN      COP       Mensual   
2           NaN             NaN      COP           NaN   

                                               title  \
0  Casa En Arriendo/venta En Bogota Sotileza Cod....   
1  SE VENDE APARTAMENTO EN EL SECTOR DE ZUÑIGA-EN...   
2           Casa Barata San Javier Solo por este mes   

                                         description  property_type  \
0  EXCLUSIVA CASA EN VENTA Y ARRIENDO, UBICADA EN...           Casa   
1  PR 7555. Apartamento Duplex en unidad cerrada,...    Apartamento   
2  Casa compraventa escrituras en \n proceso,sala...           Casa   

   operation_type                        geometry  
0           Venta           POINT (-74.075 4.722)  
1           Venta                     POINT EMPTY  
2           Venta  POINT (-75.6283806 6.25925267)  

[3 rows x 26 columns]

In [247]:
#conociendo la cantidad de columnas , 
# tipos de datos de cada columna y su cantidad de datos  de data_test 
data_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65850 entries, 0 to 65849
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       65850 non-null  int64  
 1   id               65850 non-null  object 
 2   ad_type          65850 non-null  object 
 3   start_date       65850 non-null  object 
 4   end_date         65850 non-null  object 
 5   created_on       65850 non-null  object 
 6   lat              49433 non-null  float64
 7   lon              49433 non-null  float64
 8   l1               65850 non-null  object 
 9   l2               65850 non-null  object 
 10  l3               62193 non-null  object 
 11  l4               15074 non-null  object 
 12  l5               9052 non-null   object 
 13  l6               2364 non-null   object 
 14  rooms            9110 non-null   float64
 15  bedrooms         13403 non-null  float64
 16  bathrooms        52054 non-null  float64
 17  surface_tota

In [248]:
#mostrando las 3 primeras filas  de data
data.head(3)

Unnamed: 0                        id    ad_type  start_date    end_date  \
0      861131  CScTjaFiP6yCzS8tFX0ZVA==  Propiedad  2020-12-14  9999-12-31   
1      200972  DlCxox9+FuugkxS2kBVfJg==  Propiedad  2020-09-16  2021-05-14   
2      554622  yitN1AigL45gh/K2YSF8rw==  Propiedad  2020-08-19  2020-08-20   

   created_on       lat        lon        l1            l2  ... surface_total  \
0  2020-12-14  6.203000 -75.572000  Colombia     Antioquia  ...           NaN   
1  2020-09-16  4.722748 -74.073115  Colombia  Cundinamarca  ...           NaN   
2  2020-08-19  4.709000 -74.030000  Colombia  Cundinamarca  ...           NaN   

  surface_covered         price currency  price_period  \
0             NaN  9.000000e+09      COP           NaN   
1             NaN  6.250000e+08      COP           NaN   
2             NaN  4.200000e+08      COP           NaN   

                                               title  \
0  Hermosa Mansión En El Poblado. Área Construida...   
1   APARTAMENTO EN VENTA, BOGOTA-CIUDAD JARDIN NORTE   
2  Apartamento En Venta En Bogota Bella Suiza Cod...   

                                         description  property_type  \
0  <b>Hermosa Mansión En El Poblado. Área Constru...           Casa   
1  Venta apartamento en piso 16 con vista exterio...    Apartamento   
2  Al entrar a este bonito apartamento interior e...           Casa   

   operation_type                       geometry  
0           Venta          POINT (-75.572 6.203)  
1           Venta  POINT (-74.0731146 4.7227485)  
2           Venta           POINT (-74.03 4.709)  

[3 rows x 27 columns]

In [7]:
#conociendo la cantidad de columnas , 
# tipos de datos de cada columna y su cantidad de datos de data 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197549 entries, 0 to 197548
Data columns (total 27 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Unnamed: 0       197549 non-null  int64  
 1   id               197549 non-null  object 
 2   ad_type          197549 non-null  object 
 3   start_date       197549 non-null  object 
 4   end_date         197549 non-null  object 
 5   created_on       197549 non-null  object 
 6   lat              148051 non-null  float64
 7   lon              148051 non-null  float64
 8   l1               197549 non-null  object 
 9   l2               197549 non-null  object 
 10  l3               186517 non-null  object 
 11  l4               45367 non-null   object 
 12  l5               27409 non-null   object 
 13  l6               6867 non-null    object 
 14  rooms            27537 non-null   float64
 15  bedrooms         40525 non-null   float64
 16  bathrooms        156467 non-null  floa

In [249]:
#analizando promedio de precio de la data original que es lo que 
#nos piden para poder diferenciar si las propiedas son caras o baratas
data["price"].describe()

count    1.974860e+05
mean     6.436051e+08
std      1.944868e+09
min      0.000000e+00
25%      2.000000e+08
50%      3.300000e+08
75%      6.000000e+08
max      3.450000e+11
Name: price, dtype: float64

In [250]:
#seleccionamos as columnas que consideramos más  relevantes al momento de estimar los precios
#como rooms, bedrooms, bathrooms,
#price, currency(moneda) y l1(queremos saber si todos perteneces a Colombia)
data1 = data[["rooms","bedrooms","bathrooms","price","currency","l1"]]

#verificamos si es que tanto las monedas y el país corresponden
#a pesos colombianos (COP) y Colombia, respectivamente.
data1["currency"].value_counts(), data1["l1"].value_counts()




(COP    197474
 USD         8
 Name: currency, dtype: int64,
 Colombia    197549
 Name: l1, dtype: int64)

Vemos que los resultados para data todas las 197,549 propiedades están en Colombia pero en las monedas hay 8 que corresponden a dólares, no obstante primero queremos usar una data limpia de nulos para el análisis, así que empezaremos con dropeando los nulos y luego veremos si aún se mantienen los dólares


In [251]:
#Usando el dropna nos quedamos con solamente 8907 registros y 
#vemos que ya no aparecen los dólares
data1 = data[["rooms","bedrooms","bathrooms","price","currency","l1"]].dropna()
data1["currency"].value_counts(), data1["l1"].value_counts()

(COP    8907
 Name: currency, dtype: int64,
 Colombia    8907
 Name: l1, dtype: int64)

In [252]:
#Ahora las columnas currency y l1 ya no nos sirven
#por lo que pasaremos a eliminarlas
data1 = data1[["rooms","bedrooms","bathrooms","price"]]
data1

rooms  bedrooms  bathrooms         price
42        3.0       3.0        2.0  2.400000e+08
56        3.0       3.0        3.0  4.500000e+08
83        1.0       1.0        1.0  2.780000e+08
101       4.0       4.0        4.0  5.700000e+08
102       3.0       3.0        3.0  4.980000e+08
...       ...       ...        ...           ...
197430    3.0       3.0        4.0  1.290000e+09
197458    9.0       9.0        3.0  5.000000e+08
197478   10.0      10.0       12.0  3.900000e+09
197485    3.0       3.0        3.0  2.000000e+08
197535    3.0       3.0        2.0  2.700000e+08

[8907 rows x 4 columns]

In [253]:
#creando columna precio class para determinar si es caro o barato según el promedio
data1["precio_class"] = None
data1

rooms  bedrooms  bathrooms         price precio_class
42        3.0       3.0        2.0  2.400000e+08         None
56        3.0       3.0        3.0  4.500000e+08         None
83        1.0       1.0        1.0  2.780000e+08         None
101       4.0       4.0        4.0  5.700000e+08         None
102       3.0       3.0        3.0  4.980000e+08         None
...       ...       ...        ...           ...          ...
197430    3.0       3.0        4.0  1.290000e+09         None
197458    9.0       9.0        3.0  5.000000e+08         None
197478   10.0      10.0       12.0  3.900000e+09         None
197485    3.0       3.0        3.0  2.000000e+08         None
197535    3.0       3.0        2.0  2.700000e+08         None

[8907 rows x 5 columns]

In [254]:
#usando la información de precio promedio anterior, pasamos a diferenciar
#las propiedas baratas y caras según el promedio
import numpy as np
data1["precio_class"] = np.where(data1["price"]> 6.436051e+08, "caro", "barato")
data1["precio_class"].value_counts()


barato    5997
caro      2910
Name: precio_class, dtype: int64

In [255]:
#Ahora la columna price ya no nos sirve y la eliminaremos 
data1 = data1[["rooms","bedrooms","bathrooms","precio_class"]]
data1

rooms  bedrooms  bathrooms precio_class
42        3.0       3.0        2.0       barato
56        3.0       3.0        3.0       barato
83        1.0       1.0        1.0       barato
101       4.0       4.0        4.0       barato
102       3.0       3.0        3.0       barato
...       ...       ...        ...          ...
197430    3.0       3.0        4.0         caro
197458    9.0       9.0        3.0       barato
197478   10.0      10.0       12.0         caro
197485    3.0       3.0        3.0       barato
197535    3.0       3.0        2.0       barato

[8907 rows x 4 columns]

In [256]:
#reemplazamos las variables barato por 0 y caro por 1,
#para dejarlo listo para el ML
data1 = data1.replace({'barato': 0, 'caro' : 1})
data1

rooms  bedrooms  bathrooms  precio_class
42        3.0       3.0        2.0             0
56        3.0       3.0        3.0             0
83        1.0       1.0        1.0             0
101       4.0       4.0        4.0             0
102       3.0       3.0        3.0             0
...       ...       ...        ...           ...
197430    3.0       3.0        4.0             1
197458    9.0       9.0        3.0             0
197478   10.0      10.0       12.0             1
197485    3.0       3.0        3.0             0
197535    3.0       3.0        2.0             0

[8907 rows x 4 columns]

Ahora pasaremos a ordenar nuestro data_test y lo guardaremos en un nuevo data frame llamado data2

In [257]:
#seleccionamos los datos relevantes como rooms, bedrooms, bathrooms,
#price, currency(moneda) y l1(queremos saber si todos perteneces a Colombia)
data2 = data_test[["rooms","bedrooms","bathrooms","currency","l1"]]

#verificamos si es que tanto las monedas y el país corresponden
#a pesos colombianos (COP) y Colombia, respectivamente.
data2["currency"].value_counts(), data2["l1"].value_counts()

(COP    65832
 USD        1
 Name: currency, dtype: int64,
 Colombia    65850
 Name: l1, dtype: int64)

Vemos que los resultados para data2 todas las 65,850 propiedades están en Colombia pero en las monedas hay 1 que corresponden a dólares, no obstante primero queremos usar una data limpia de nulos para el análisis, así que empezaremos con dropeando los nulos y luego veremos si aún se mantienen los dólares

In [258]:
#Usando el dropna nos quedamos con solamente 8907 registros y 
#vemos que ya no aparecen los dólares
data2 = data2[["rooms","bedrooms","bathrooms","currency","l1"]].dropna()
data2["currency"].value_counts(), data2["l1"].value_counts()

(COP    2979
 Name: currency, dtype: int64,
 Colombia    2979
 Name: l1, dtype: int64)

In [259]:
#Ahora las columnas currency y l1 ya no nos sirven
#por lo que pasaremos a eliminarlas
data2 = data2[["rooms","bedrooms","bathrooms"]]
data2

rooms  bedrooms  bathrooms
0        3.0       3.0        3.0
14       3.0       3.0        3.0
26       9.0       9.0       10.0
52       1.0       1.0        1.0
112      3.0       3.0        2.0
...      ...       ...        ...
65740    3.0       3.0        2.0
65775    3.0       3.0        2.0
65779    2.0       2.0        2.0
65804    2.0       2.0        2.0
65824    6.0       6.0        3.0

[2979 rows x 3 columns]

DATA1 = DATATRAIN LIMPIO;
DATA2 = DATATEST LIMPIO

In [60]:
#Vamos a usar RandomForestClassifier, que es bosque aleatorio 
#lo cual nos va permitir resultados más precisos
from sklearn.ensemble import RandomForestClassifier


n_estimators = cantidad de árboles , se considerará 100 árboles  ; 
max_features = cantidad de características para cada árbol, se úsara sqrt, la raíz cuadrada    ;
bootstrap = muestreo aleatorio de datos , le damos True, aunque viene por defecto      ;
max_samples = procentaje de muestreo aleatorio, usaremos 2/3
oob_score = son los datos que quedaron fuera de los samples

In [104]:
rf = RandomForestClassifier(n_estimators = 100, 
                            max_features = "sqrt",
                            bootstrap=True, 
                            max_samples= 2/3, 
                            oob_score= True)
data1_f = data1.drop(['precio_class'], axis=1)
rf.fit(data1_f, data1["precio_class"])

RandomForestClassifier(max_samples=0.6666666666666666, oob_score=True)

In [145]:
#predict lo usamos para predecir y usamos la data2
print(rf.predict(data2))
pred = rf.predict(data2)
print(rf.score(data1_f, data1["precio_class"]))
print(rf.oob_score_)

[0 0 1 ... 0 0 0]
0.8049848433816099
0.7992590097675986


In [109]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import recall_score

In [175]:
pred_pd = pd.DataFrame(pred)
#a = confusion_matrix(data1, data1[["precio_class"]])
type(pred_pd)

pandas.core.frame.DataFrame

In [138]:
b = np.array(data2)

In [148]:
a = confusion_matrix(pred, pred )
a

array([[2186,    0],
       [   0,  793]], dtype=int64)

In [151]:
matriz = data2
matriz , pred_pd

(       rooms  bedrooms  bathrooms
 0        3.0       3.0        3.0
 14       3.0       3.0        3.0
 26       9.0       9.0       10.0
 52       1.0       1.0        1.0
 112      3.0       3.0        2.0
 ...      ...       ...        ...
 65740    3.0       3.0        2.0
 65775    3.0       3.0        2.0
 65779    2.0       2.0        2.0
 65804    2.0       2.0        2.0
 65824    6.0       6.0        3.0
 
 [2979 rows x 3 columns],
 0       0
 1       0
 2       1
 3       0
 4       0
        ..
 2974    0
 2975    0
 2976    0
 2977    0
 2978    0
 Length: 2979, dtype: int64)

In [267]:
m=data2
m


rooms  bedrooms  bathrooms  m
0        3.0       3.0        3.0  0
14       3.0       3.0        3.0  0
26       9.0       9.0       10.0  1
52       1.0       1.0        1.0  0
112      3.0       3.0        2.0  0
...      ...       ...        ... ..
65740    3.0       3.0        2.0  0
65775    3.0       3.0        2.0  0
65779    2.0       2.0        2.0  0
65804    2.0       2.0        2.0  0
65824    6.0       6.0        3.0  0

[2979 rows x 4 columns]

In [266]:
pred_pd

0
0     0
1     0
2     1
3     0
4     0
...  ..
2974  0
2975  0
2976  0
2977  0
2978  0

[2979 rows x 1 columns]

In [268]:
#m = m.insert(3,"m", pred ,True)
m

rooms  bedrooms  bathrooms  m
0        3.0       3.0        3.0  0
14       3.0       3.0        3.0  0
26       9.0       9.0       10.0  1
52       1.0       1.0        1.0  0
112      3.0       3.0        2.0  0
...      ...       ...        ... ..
65740    3.0       3.0        2.0  0
65775    3.0       3.0        2.0  0
65779    2.0       2.0        2.0  0
65804    2.0       2.0        2.0  0
65824    6.0       6.0        3.0  0

[2979 rows x 4 columns]

In [286]:
#from sklearn.model_selection import train_test_split
m

rooms  bedrooms  bathrooms  m
0        3.0       3.0        3.0  0
14       3.0       3.0        3.0  0
26       9.0       9.0       10.0  1
52       1.0       1.0        1.0  0
112      3.0       3.0        2.0  0
...      ...       ...        ... ..
65740    3.0       3.0        2.0  0
65775    3.0       3.0        2.0  0
65779    2.0       2.0        2.0  0
65804    2.0       2.0        2.0  0
65824    6.0       6.0        3.0  0

[2979 rows x 4 columns]

In [296]:
X= m.drop(columns=["m"])
y= m.m
X_train, X_test, y_train, y_test = train_test_split(X ,y, test_size=1, random_state=42, stratify=y)

ValueError: The test_size = 1 should be greater or equal to the number of classes = 2

In [298]:
cm= confusion_matrix(m.m, pred)

In [299]:
cm

array([[2186,    0],
       [   0,  793]], dtype=int64)

FIN

FIN

In [107]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197549 entries, 0 to 197548
Data columns (total 27 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Unnamed: 0       197549 non-null  int64  
 1   id               197549 non-null  object 
 2   ad_type          197549 non-null  object 
 3   start_date       197549 non-null  object 
 4   end_date         197549 non-null  object 
 5   created_on       197549 non-null  object 
 6   lat              148051 non-null  float64
 7   lon              148051 non-null  float64
 8   l1               197549 non-null  object 
 9   l2               197549 non-null  object 
 10  l3               186517 non-null  object 
 11  l4               45367 non-null   object 
 12  l5               27409 non-null   object 
 13  l6               6867 non-null    object 
 14  rooms            27537 non-null   float64
 15  bedrooms         40525 non-null   float64
 16  bathrooms        156467 non-null  floa

In [18]:
data.head(3)

Unnamed: 0                        id    ad_type  start_date    end_date  \
0      861131  CScTjaFiP6yCzS8tFX0ZVA==  Propiedad  2020-12-14  9999-12-31   
1      200972  DlCxox9+FuugkxS2kBVfJg==  Propiedad  2020-09-16  2021-05-14   
2      554622  yitN1AigL45gh/K2YSF8rw==  Propiedad  2020-08-19  2020-08-20   

   created_on       lat        lon        l1            l2  ... surface_total  \
0  2020-12-14  6.203000 -75.572000  Colombia     Antioquia  ...           NaN   
1  2020-09-16  4.722748 -74.073115  Colombia  Cundinamarca  ...           NaN   
2  2020-08-19  4.709000 -74.030000  Colombia  Cundinamarca  ...           NaN   

  surface_covered         price currency  price_period  \
0             NaN  9.000000e+09      COP           NaN   
1             NaN  6.250000e+08      COP           NaN   
2             NaN  4.200000e+08      COP           NaN   

                                               title  \
0  Hermosa Mansión En El Poblado. Área Construida...   
1   APARTAMENTO EN VENTA, BOGOTA-CIUDAD JARDIN NORTE   
2  Apartamento En Venta En Bogota Bella Suiza Cod...   

                                         description  property_type  \
0  <b>Hermosa Mansión En El Poblado. Área Constru...           Casa   
1  Venta apartamento en piso 16 con vista exterio...    Apartamento   
2  Al entrar a este bonito apartamento interior e...           Casa   

   operation_type                       geometry  
0          Compra          POINT (-75.572 6.203)  
1          Compra  POINT (-74.0731146 4.7227485)  
2          Compra           POINT (-74.03 4.709)  

[3 rows x 27 columns]

In [14]:
print(data["operation_type"] == "Compra")

0         True
1         True
2         True
3         True
4         True
          ... 
197544    True
197545    True
197546    True
197547    True
197548    True
Name: operation_type, Length: 197549, dtype: bool


In [16]:
data["operation_type"].describe()

count     197549
unique         1
top       Compra
freq      197549
Name: operation_type, dtype: object